git/git_cached_repository.bzl

```
@@ -0,0 +1,222 @@
+load("//git:git_cached_worker.bzl", "checkout", "clone","fetch")
+
+"""Rules for caching external git repositories in favor of keeping a lighter git network overhead."""
+
+load(
+    "@bazel_tools//tools/build_defs/repo:utils.bzl",
+    "patch",
+    "update_attrs",
+    "workspace_and_buildfile",
+)
+
+_env_var_enable_verbosity = "GIT_CACHED_VERBOSE"
+
+_common_attrs = {
+    "remote_url":
+        attr.string(
+            mandatory=True,
+            doc="The URI of the remote Git repository",
+        ),
+    "commit":
+        attr.string(
+            mandatory=True,
+            doc="Specific commit to be checked out",
+        ),
+    "branch":
+        attr.string(
+            default="master",
+            doc="Specify a branch to fetch HEAD references",
+        ),
+    "shallow_since":
+        attr.string(
+            default="",
+            doc="An optional date in addition to a specified commit. " +
+            "Setting such a date close to the " +
+            "specified commit allows for a more shallow clone of the " +
+            "repository, saving bandwidth and wall-clock time.",
+        ),
+    "cache_directory":
+        attr.string(
+            default="{}/.git-cache/{}",
+            doc="Local path to clone the repository into, " +
+            "a default location can be used to clone into $HOME/.git-cache/<repo-name>",
+        ),
+    "fetch_retries_count":
+        attr.int(
+            default=2,
+            doc=
+            "Amount of retries attempts for fetching a repository before failing",
+        ),
+    "fetch_retry_timeout_in_sec":
+        attr.int(
+            default=300,
+            doc="Time to wait for a single fetch retry attempt",
+        ),
+}
+
+
+def _is_verbosity_enabled(repo_ctx):
+    return repo_ctx.os.environ.get(_env_var_enable_verbosity, "") == "True"
+
+
+def _log(repo_ctx, message, report_progress=False):
+    if report_progress:
+        repo_ctx.report_progress(message)
+
+    if _is_verbosity_enabled(repo_ctx):
+        print(message)
+
+
+def _get_local_cache_repo_path(repo_ctx):
+    """ Return a default local path to clone the repository into else return the cache_directory attribute if supplied
+        (default location: $HOME/.git-cache/<repo-name>)
+
+        Args:
+            repo_ctx: Repository context of the calling rule
+
+        Returns:
+            Path to the local repository cache folder
+    """
+    if "{}/.git-cache/{}" in repo_ctx.attr.cache_directory:
```

it feels kinda awkward that this is the default value, maybe it's better to actually set default value to `None` and then this line would look like:
```suggestion
    if not repo_ctx.attr.cache_directory:
```

---

Set `cache_directory ` default to `None` and using `{}/.git-cache/{}` path explicitly if no path supplied.